File: Problem2.py
'''

Time Complexity : O(n)
Space Complexity : O(n)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No issues faced

Your code here along with comments explaining your approach:
Firstly look for the object in the list provided and append it into the queue the object which matches the id. Then subsequently add up the
importance using BFS 

'''

class Solution:
    def getImportance(self, employees: List['Employee'], id: int) -> int:
        q = collections.deque()

        for i in employees:
            if i.id == id:
                q.append(i)


        imp = 0
        while q:
            emp = q.popleft()
            imp += emp.importance

            for a in emp.subordinates:
                q.append(next(e for e in employees if e.id == a))

                # next returns the first element in the iteration

        return imp

File: problem2.cpp
// TC: O(N);
// SC: O(N)
// Did it run on leetcode: yes


// we take a map (key as id and value as Employee*) to ease our searching. then we use dfs/bfs
/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    unordered_map<int,Employee*>map;
    int ans = 0;
    int getImportance(vector<Employee*> employees, int id) {
        for(int i = 0;i< employees.size();i++)
        {
            map.insert({employees[i]->id, employees[i]});
        }
        dfs(id);
        return ans;
    }
    void dfs(int id)
    {
        Employee* curr = map[id];
        
        ans+= curr->importance;
        vector<int>v = curr->subordinates;
        for(int i = 0;i< v.size();i++)
        {
            dfs(v[i]);
        }
    }
};



/*
// Definition for Employee.
class Employee {
public:
    int id;
    int importance;
    vector<int> subordinates;
};
*/

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        unordered_map<int,Employee*>map;
        for(int i = 0;i<employees.size();i++)
        {
            map.insert({employees[i]->id, employees[i]});
        }
        queue<int>q;
        q.push(id);
        int ans=0;
        while(!q.empty())
        {
            Employee* curr = map[q.front()];q.pop();
            ans+= curr->importance;
            vector<int>v = curr->subordinates;
            for(int i = 0;i< v.size();i++)
            {
                q.push(v[i]);
            }
        }
        return ans;
    }
};